### ISSUE TITLE
How to instrument very similar modules?

### ISSUE THREAD

**Issue by samreid** [phetsims/together#28]
I wanted to try to instrument EnergySkatePark:Basics so I could investigate how to instrument very similar screens (though the same issues will apply to other kinds of similar modules as well).

I first tried a "brute force" approach, hardcoding  componentID declarations and pushing them back as far as necessary to reach all of the components: 61513e622b03219f4aa83b642341ca6d8ff0006d

This solutions has the benefit that componentID strings always appear as whole strings and are hence easily searchable, but there is a lot of boilerplate and duplication which could lead to errors.

Perhaps we should experiment with a more dynamic approach, such as passing the screen name through and using that to concatenate componentIDs.  This could save on having to pass explicit componentIDs through everywhere at the cost of making it more difficult to search/find the componentIDs in the code.

Another possibility that crossed my mind was to pass the screen componentID map through to every component (unmodified) and let each component draw its componentID from that map.  That would be somewhat similar to passing through the screen name for concatenation, but may expose too much information (TMI) to all components.

**Comment by samreid** [comment]
Here is a better link to the ESPB sha for discussion:
https://github.com/phetsims/energy-skate-park-basics/commit/61513e622b03219f4aa83b642341ca6d8ff0006d

Maybe a good generalization would be to pass through a contextID which can be prepended.  Then this solution can be used for other types of generalizations, not just duplicated across screens.

**Comment by samreid** [comment]
There is the parallel question of if/how to productively avoid duplicating information within the API file itself.

**Comment by samreid** [comment]
Perhaps if we use a contextID, it should be an object with an append method, so it will be easy to search for usages of this pattern, and guarantee consistent behavior (factor out x +'.'+y at the very least).  Not sure where such a thing should live--we may eventually need a MIT licensed support library for together for other features such as handling Array element id assignment.

**Comment by samreid** [comment]
Brainstorming names for a MIT licensed support library for together features that must appear in published sims:

tandem

Any other ideas?  I may start with tandem and see how it goes.

**Comment by samreid** [comment]
The above commit "Moved gridCheckBox to ComponentIDContext..." is a good example of the tradeoff between the brute force and dynamic (context-based) approaches.  Seems like a good direction to move in--the brute force approach requires too much manual labor and has too many opportunities for typos.

**Comment by samreid** [comment]
With additional change sets not recorded here, I've finished moving ESPB to ComponentIDContext, and it is looking much better.  One other tradeoff I noticed is this decreases flexibility for giving individual component id's.  Perhaps a good reason to use this pattern in sim code but not in common code?  If common code used ComponentIDContext, then a child component name would be fixed by common code and the simulation would no longer be able to flexibly give it any componentID--maybe this is a good thing, but maybe we would need to circumvent that convention in some cases.
